### PR TITLE
Computing next time-step size from CFL

### PR DESCRIPTION
**Pull request by jafranc**
This PR is aiming at having the following time-step size being calculated so that it honors `CFL< m_cflFactor`. The max value is set for phase or component CFL, which ever is the largest.
`m_cflFactor` is already member of `SoverBase`.

Proof of Concept Stage
--------------------------------
~Its default value has been changed to _-1_ on order to easily detect enable/disable.~
~Its only found use in `SurfaceGenerator` has been modified so that it still honors its former default value of `.5`.~

Long term Stage
-----------------------
- Prior version is restored.
- `ComputationalMultiphaseBase` is holding its own `m_targetCFLFlow` not interfering with any prior implementation. 
- CFL call is refactored to be called as a time-step size estimator and for statistics.
- One integrated test is altered to use the feature `4comp_2ph_1d`

Rebaseline https://github.com/GEOS-DEV/integratedTests/pull/72

### CODE REVIEW

File: src/coreComponents/physicsSolvers/SolverBase.cpp [hunk 1]
```diff
@@ -52,7 +52,7 @@ SolverBase::SolverBase( string const & name,
   this->setTimesteppingBehavior( ExecutableGroup::TimesteppingBehavior::DeterminesTimeStepSize );
 
   registerWrapper( viewKeyStruct::cflFactorString(), &m_cflFactor ).
-    setApplyDefaultValue( 0.5 ).
+    setApplyDefaultValue( -1.0 ).
     setInputFlag( InputFlags::OPTIONAL ).
     setDescription( "Factor to apply to the `CFL condition <http://en.wikipedia.org/wiki/Courant-Friedrichs-Lewy_condition>`_"
                     " when calculating the maximum allowable time step. Values should be in the interval (0,1] " );
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/SolverBase.cpp:58]: for FIM, does it still have to be <=1? What if one wants to push CFL to 10, 100, etc?

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp [hunk 2]
```diff
@@ -1928,6 +1945,163 @@ real64 CompositionalMultiphaseBase::setNextDtBasedOnStateChange( real64 const &
   return std::min( std::min( nextDtPressure, nextDtPhaseVolFrac ), nextDtTemperature );
 }
 
+real64 CompositionalMultiphaseBase::setNextDtBasedOnCFL( const geos::real64 & currentDt, geos::DomainPartition & domain )
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:1948]: this duplicates a lot from CompositionalMultiphaseStatistics::computeCFLNumbers, right? any way to share the code between the two?

**Review by paveltomin — APPROVED** [review]
overall looks good

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp (hunk 2 above)
Comment (jafranc) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:1975]: Will do some clean up

File: src/coreComponents/physicsSolvers/SolverBase.cpp (hunk 1 above)
Comment (jafranc) [reply src/coreComponents/physicsSolvers/SolverBase.cpp:59]: We can push it above 1 for sure. For instance to test C1-ppu and IHU I ran cases at 50 and 320 as reported in articles. `0.5` is more as an obvious default than a recommendation

Comment (paveltomin) [reply src/coreComponents/physicsSolvers/SolverBase.cpp:59]: Ok great, maybe change the description to avoid confusion?

Comment (jafranc) [reply src/coreComponents/physicsSolvers/SolverBase.cpp:59]: Ok, I will the introduce a new _CFL_ for flow, as I was, during the development, living off the `m_cflFactor` that is being used in hydrofrac to spare me some proliferation of CFLs. At this point I think it is legit to make difference between those two thing.

File: src/coreComponents/physicsSolvers/SolverBase.cpp [hunk 3]
```diff
@@ -290,32 +290,34 @@ real64 SolverBase::setNextDt( real64 const & currentDt,
...
     }
   }
-  else // time step size decided based on state change
+  else         // time step size decided based on state change
   {
     if( nextDtStateChange > currentDt )
     {
       GEOS_LOG_LEVEL_RANK_0( 1, GEOS_FMT( "{}: Time-step required will be increased based on state change.",
-                                          getName() ) );
+                                          getName()));
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/SolverBase.cpp:317]: how is uncrustify not cleaning things like that? @tbeltzun would it after your PR?

File: src/coreComponents/physicsSolvers/SolverBase.cpp [hunk 4]
```diff
@@ -353,6 +355,15 @@ real64 SolverBase::setNextDtBasedOnNewtonIter( real64 const & currentDt )
   return nextDt;
 }
 
+
+real64 SolverBase::setNextDtBasedOnCFL( const geos::real64 & currentDt, geos::DomainPartition & domain )
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/SolverBase.cpp:361]: it's ok for this PR but maybe it's time to generalize it into `setNextDt` that loops through different criteria

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp (hunk 2 above)
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:2083]: looks good, thanks for cleaning it up

File: src/coreComponents/physicsSolvers/surfaceGeneration/SurfaceGenerator.cpp [hunk 5]
```diff
@@ -726,8 +726,9 @@ int SurfaceGenerator::separationDriver( DomainPartition & domain,
   real64 ruptureRate = calculateRuptureRate( elementManager.getRegion< SurfaceElementRegion >( this->m_fractureRegionName ) );
 
   GEOS_LOG_LEVEL_RANK_0( 3, "rupture rate is " << ruptureRate );
-  if( ruptureRate > 0 )
-    m_nextDt = ruptureRate < 1e99 ? m_cflFactor / ruptureRate : 1e99;
+  if( ruptureRate > 0 ) //FIXME ugly adaptation while waiting to know if we have to have a new spec flow value or can factorize.
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/surfaceGeneration/SurfaceGenerator.cpp]: @jafranc what is going on here? :)

File: src/coreComponents/schema/schema.xsd [hunk 6]
```diff
@@ -2008,7 +2008,7 @@ the relative residual norm satisfies:
 			<xsd:element name="NonlinearSolverParameters" type="NonlinearSolverParametersType" maxOccurs="1" />
 		</xsd:choice>
 		<!--cflFactor => Factor to apply to the `CFL condition <http://en.wikipedia.org/wiki/Courant-Friedrichs-Lewy_condition>`_ when calculating the maximum allowable time step. Values should be in the interval (0,1] -->
-		<xsd:attribute name="cflFactor" type="real64" default="0.5" />
+		<xsd:attribute name="cflFactor" type="real64" default="-1" />
```
Comment (paveltomin) [inline src/coreComponents/schema/schema.xsd]: should it be committed?

**Review by paveltomin — APPROVED** [review]
looks good
_wish uncrustify to be more consistent about spaces and new lines_

File: src/coreComponents/physicsSolvers/SolverBase.cpp (hunk 4 above)
Comment (jafranc) [reply src/coreComponents/physicsSolvers/SolverBase.cpp:361]: Agreed, though this one is a bit specific as we want to force dt on being tracking CFL. But 'timestepping' deserves its own more elaborate structures IMO as it is way more complex.

File: src/coreComponents/physicsSolvers/surfaceGeneration/SurfaceGenerator.cpp (hunk 5 above)
Comment (jafranc) [reply src/coreComponents/physicsSolvers/surfaceGeneration/SurfaceGenerator.cpp]: lol, I restored the `develop` version I was highjacking but forgot to erase my reminder :p

File: src/coreComponents/physicsSolvers/SolverBase.cpp (hunk 3 above)
Comment (tbeltzun) [reply src/coreComponents/physicsSolvers/SolverBase.cpp:317]: It looks like uncrustify doesn't enforce spacing around parentheses, when they span over multiple lines.

**Comment by paveltomin** [comment]
@jafranc clean up and merge into the queue?

**Comment by jafranc** [comment]
> @jafranc clean up and merge into the queue?

Will do soon. IIRC it passed all CPU tests. Let us see GPU. Then if someone want to rebaseline it on test `4comp_2ph_1d.xml`.

File: src/coreComponents/schema/schema.xsd (hunk 6 above)
Comment (jafranc) [reply src/coreComponents/schema/schema.xsd]: No, you're right, just revert it in a later commit (hopefully)

**Comment by jafranc** [comment]
Note the rebaselines present in the `integratedTests` PRs are outdated !

**Comment by TotoGaz** [comment]
> Note the rebaselines present in the `integratedTests` PRs are outdated !

Just ditch them, we'll do fresh ones?

**Comment by TotoGaz** [comment]
Hey @jafranc can you manage the conflict so we can merge your PR?

**Comment by rrsettgast** [comment]
@jafranc Is there an integratedTests PR that has the baselines?

**Comment by jafranc** [comment]
> @jafranc Is there an integratedTests PR that has the baselines?

No, mine was outdated, so as suggested by @TotoGaz I discarded it.
From the rebasline test, it seems to be consistent with the added fields and diffs in `4comp_2ph_1d` as it has been changed to use the new feature.

 Note that `PoroElastic` has diff as already noticed by @TotoGaz

**Review by jafranc — COMMENTED** [review]
Thank you @TotoGaz

**Comment by TotoGaz** [comment]
@CusiniM I did not manage to rebaseline this PR.
I had issues with the new restart (?) steps of `4comp_2ph_1d` cases, sometimes with something like `/tmp/geos/integratedTests/tests/allTests/compositionalMultiphaseFlow/4comp_2ph_1d_01/0to218_restart_000000011/rank_0000000.hdf5 and /tmp/geos/integratedTests/tests/allTests/compositionalMultiphaseFlow/baselines/4comp_2ph_1d_01/0to218_restart_000000218/rank_0000000.hdf5`
I've rolled back everything. Can you give it a shot since you're more experienced than me on this kind of stuff?

**Comment by jafranc** [comment]
> @CusiniM I did not manage to rebaseline this PR. I had issues with the new restart (?) steps of `4comp_2ph_1d` cases, sometimes with something like `/tmp/geos/integratedTests/tests/allTests/compositionalMultiphaseFlow/4comp_2ph_1d_01/0to218_restart_000000011/rank_0000000.hdf5 and /tmp/geos/integratedTests/tests/allTests/compositionalMultiphaseFlow/baselines/4comp_2ph_1d_01/0to218_restart_000000218/rank_0000000.hdf5` I've rolled back everything. Can you give it a shot since you're more experienced than me on this kind of stuff?

Let me know if you want me to do it. As the time-steps are now decided upon CFL and not any other heuristic, this might introduce diffuculties on restart and we might want to force restart time-step one way or another.

**Comment by TotoGaz** [comment]
Yes @jafranc surely there's something to do with the restarts https://github.com/GEOS-DEV/integratedTests/blob/develop/tests/allTests/compositionalMultiphaseFlow/compositionalMultiphaseFlow.ats
Can you have a look and keep us posted?
Meanwhile I'll put your PR back in the `Testing` column, we'll put it back to the `Merge Queue` asap.

**Comment by paveltomin** [comment]
> Yes @jafranc surely there's something to do with the restarts https://github.com/GEOS-DEV/integratedTests/blob/develop/tests/allTests/compositionalMultiphaseFlow/compositionalMultiphaseFlow.ats Can you have a look and keep us posted? Meanwhile I'll put your PR back in the `Testing` column, we'll put it back to the `Merge Queue` asap.

`check_step=218` for `4comp_2ph_1d` needs to be adjusted to match current last time step number, should be relatively straightforward, let me check...

**Comment by jafranc** [comment]
> > Yes @jafranc surely there's something to do with the restarts https://github.com/GEOS-DEV/integratedTests/blob/develop/tests/allTests/compositionalMultiphaseFlow/compositionalMultiphaseFlow.ats Can you have a look and keep us posted? Meanwhile I'll put your PR back in the `Testing` column, we'll put it back to the `Merge Queue` asap.
> 
> `check_step=218` for `4comp_2ph_1d` needs to be adjusted to match current last time step number, should be relatively straightforward, let me check...

Yes, that was what I did I think during my first (outdated) rebaseline.

**Comment by paveltomin** [comment]
> > > Yes @jafranc surely there's something to do with the restarts https://github.com/GEOS-DEV/integratedTests/blob/develop/tests/allTests/compositionalMultiphaseFlow/compositionalMultiphaseFlow.ats Can you have a look and keep us posted? Meanwhile I'll put your PR back in the `Testing` column, we'll put it back to the `Merge Queue` asap.
> > 
> > 
> > `check_step=218` for `4comp_2ph_1d` needs to be adjusted to match current last time step number, should be relatively straightforward, let me check...
> 
> Yes, that was what I did I think during my first (outdated) rebaseline.

can someone please try this `d3347c1cd0e4f7606af8aa1cd147ec51e9df4f58` hash for integratedTests submodule?
(i am in the middle of some other branch development...)

**Comment by paveltomin** [comment]
@TotoGaz @jafranc should be good to merge now

**Comment by jafranc** [comment]
> @TotoGaz @jafranc should be good to merge now

Works on my side too. Many thanks @paveltomin

**Pull request merged**